core/src/main/scala/caliban/interop/jsoniter/jsoniter.scala

```diff
@@ -78,20 +80,28 @@ private[caliban] object ValueJsoniter {
     out.writeObjectEnd()
   }
 
+  /**
+   * Note on performance:
+   *
+   * Keep number of cases at 12 (+1 default) or less, and avoid using unapply methods in the pattern-matching.
+   * This number was manually determined based on trial-and-error and was found that adding more cases reduces performance
+   * noticeably.
+   */
   private def encodeResponseValue(x: ResponseValue, out: JsonWriter, depth: Int): Unit = x match {
-    case StringValue(value)                 => out.writeVal(value)
-    case BooleanValue(value)                => out.writeVal(value)
-    case IntValue.IntNumber(value)          => out.writeVal(value)
-    case IntValue.LongNumber(value)         => out.writeVal(value)
-    case IntValue.BigIntNumber(value)       => out.writeVal(value)
-    case FloatValue.FloatNumber(value)      => out.writeVal(value)
-    case FloatValue.DoubleNumber(value)     => out.writeVal(value)
-    case FloatValue.BigDecimalNumber(value) => out.writeVal(value)
-    case NullValue                          => out.writeNull()
-    case EnumValue(value)                   => out.writeVal(value)
-    case ResponseValue.ObjectValue(o)       => writeResponseObject(o, out, decrDepth(depth, out))
-    case ResponseValue.ListValue(l)         => writeResponseArray(l, out, decrDepth(depth, out))
-    case s: ResponseValue.StreamValue       => out.writeVal(s.toString)
+    case v: StringValue                 => out.writeVal(v.value)
+    case v: BooleanValue                => out.writeVal(v.value)
+    case v: IntValue.IntNumber          => out.writeVal(v.value)
+    case NullValue                      => out.writeNull()
+    case v: EnumValue                   => out.writeVal(v.value)
+    case v: ResponseValue.ObjectValue   => writeResponseObject(v.fields, out, decrDepth(depth, out))
+    case v: ResponseValue.ListValue     => writeResponseArray(v.values, out, decrDepth(depth, out))
+    // Not very common, so we leave them last
+    case v: IntValue.LongNumber         => out.writeVal(v.value)
+    case v: IntValue.BigIntNumber       => out.writeVal(v.value)
+    case v: FloatValue.FloatNumber      => out.writeVal(v.value)
+    case v: FloatValue.DoubleNumber     => out.writeVal(v.value)
+    case v: FloatValue.BigDecimalNumber => out.writeVal(v.value)
+    case v                              => out.writeVal(v.toString)
```

Comment:
This is going to bite us some day 🥲 
Couldn't we group the remaining `IntValue` (and `FloatValue`) into one case and do a second pattern matching?

Reply:
I was thinking the same thing, but then I thought it's very unlikely we'll add more ResponseValue types. But the more I think of it, I think you're right, better be safe than sorry.

Reply:
I tried a couple of different ways in order to fully preserve type-safety, and the only feasible way to do it was to have a "full encoder" to fallback to. Encoding non-common types won't be too great, but at least we get really good performance when for common types, and we get to preserve type-safety

Reply:
What's the issue with 
```scala
case v: IntValue      => some method that only pattern match on IntValue
case v: FloatValue  => some method that only pattern match on FloatValue
```
at the end and no `case _ =>`?
We can still keep `IntNumber` at the top.

Reply:
So the issue with `IntValue` is that both `IntNumber` and `LongNumber` are commonly used (`LongNumber` is used a fair bit in some wrappers like `ApolloTracing`), so handling it generically doesn't make sense cause the only case left is `BigIntNumber`.

I ended up handling only `FloatValue` generically, while leaving `DoubleNumber` in the main pattern match cause I think that's something that's somewhat commonly used. Running the benchmarks shows that we're still OK with the number of cases that we have this way